cannon/mipsevm/memory/binary_tree.go

```diff
@@ -0,0 +1,121 @@
+package memory
+
+import (
+	"math/bits"
+)
+
+// BinaryTreeIndex is a representation of the state of the memory in a binary merkle tree.
+type BinaryTreeIndex struct {
+	// generalized index -> merkle root or nil if invalidated
+	nodes map[uint64]*[32]byte
+	// Reference to the page table from Memory.
+	pageTable map[Word]*CachedPage
+}
+
+func NewBinaryTreeMemory() *Memory {
+	pages := make(map[Word]*CachedPage)
+	index := NewBinaryTreeIndex()
+	index.setPageBacking(pages)
+	return &Memory{
+		merkleIndex:  index,
+		pageTable:    pages,
+		lastPageKeys: [2]Word{^Word(0), ^Word(0)}, // default to invalid keys, to not match any pages
+	}
+}
+
+func NewBinaryTreeIndex() *BinaryTreeIndex {
+	return &BinaryTreeIndex{
+		nodes: make(map[uint64]*[32]byte),
+	}
+}
+func (m *BinaryTreeIndex) setPageBacking(pages map[Word]*CachedPage) {
+	m.pageTable = pages
+}
+
+func (m *BinaryTreeIndex) New() PageIndex {
```

Comment:
This could accept the `pages` map and set it on the new instance. Then I think `setPageBacking` could be removed from the `PageIndex` interface.